shared/src/logs.rs

```diff
@@ -0,0 +1,302 @@
+use pgrx::*;
+use serde::{Deserialize, Serialize};
+use std::fmt::{Display, Formatter};
+
+#[allow(dead_code)]
+pub const DEFAULT_LOG_LEVEL: LogLevel = LogLevel::INFO;
+
+// Logs will live in the table created below.
+// The schema must already exist when this code is executed.
+extension_sql!(
+    r#"
+    CREATE TABLE IF NOT EXISTS paradedb.logs (
+        id SERIAL PRIMARY KEY,
+        timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+        level TEXT NOT NULL,
+        module TEXT NOT NULL,
+        file TEXT NOT NULL,
+        line INTEGER NOT NULL,
+        message TEXT NOT NULL,
+        json JSON,
+        pid INTEGER NOT NULL,
+        backtrace TEXT
+    );
+    "#
+    name = "create_paradedb_logs_table"
+);
+
+/// A logging macro designed for use within the ParadeDB system. It facilitates logging
+/// messages at various levels, optionally including additional JSON data. This macro supports
+/// three forms of invocation, allowing for flexibility in log detail.
+///
+/// # Forms
+///
+/// 1. Basic Logging: `plog!($msg:expr)`
+///    Logs a message using the default log level.
+/// 2. Logging with Additional JSON Data: `plog!($msg:expr, $json:expr)`
+///    Logs a message along with additional JSON data using the default log level.
+/// 3. Logging with Specified Level and JSON Data: `plog!($level:expr, $msg:expr, $json:expr)`
+///    Logs a message with a specified log level and additional JSON data.
+///    Accepts any type that implements Serialize.
+///
+/// # Examples
+///
+/// Basic Logging:
+/// ```no_run
```

Comment:
what is `no_run`?

Reply:
I've removed this, but happy to answer your question, because it took some research on my end as well.

Rust has a very cool feature called doctests, which will actually execute codeblocks in your documentation during `cargo test`. This helps ensure documentation doesn't go stale and show broken examples. 

However, it gets complicated with macro documentation. The doctests will be fail even though the code works... because you need some extra magic syntax for macros. I was attempting with this `no_run` to try and get this code block ignored by the test runner, but it didn't work. Leaving it here was an accident.

I spent a little time on it, but decided that it would be better to just turn doctests off for `shared`. You'll see the new line in `shared/Cargo.toml`.

`shared` is an internal library anyways, so our contracts with exported functions are already tested with Rust's type system. Our code won't even compile if contracts in `shared ` change. The doctests are more important for externally-facing libraries, where users rely on your documentation to figure out how to use your code.

I think we should make full use of doctests in `pg_bm25` and our other extensions. But turning them off in `shared` seemed like the easiest way to avoid headaches.